### ISSUE TITLE
Strange behaviour on DTLS handshake with fork() - odd clients timeout and server throws no error.

### ISSUE THREAD

**Issue by WhiteaglePT** [ARMmbed/mbedtls#2720]
### Description
- Type: Question / Bug?
- Priority: Blocker

---------------------------------------------------------------
## Bug / Question

Hi,

I'm trying to fork a DTLS server in order to be able to have multiple connections within one solution.

I've been playing with mbedtls for about a week and I've been able to do amazing things with it, however I'm currently in need of transforming my TLS client to DTLS and will need a DTLS server.

I'm trying to fork() the connection and I managed to make it work, however I'm facing the strangest behaviour:
Only **pair** clients are able to connect. I've been trying to track this down for a good 48 hours with no success.

So, the first client connects just fine,
I connect a second client and it gets stuck on the handshake.

However, if I connect a third client right after it works, and so on.

Every client that connects after a failure works, every client that connects after a successful connection doesn't!

There's no error on the server side, just a timeout on the client side and consecutive retries - all end up in error.

Now the strange part:
The good clients hello goes as expected (linked bellow "good.txt", redacted), but every odd client (linked bellow "bad.txt", original logs) freaks out and even the logs show impossible things, as they're missing some debug lines and showing others just a couple lines bellow...
(Logs bellow made with debug level 5)
[good.txt](https://github.com/ARMmbed/mbedtls/files/3318654/good.txt)
[bad.txt](https://github.com/ARMmbed/mbedtls/files/3318653/bad.txt)

I've tried resetting and/or reinitializing almost everything in group and separate after a successful connection on the main thread, I've tried to track the library code but without any success... no matter what I do **only pair clients work** and the logs come back nearly blank!

Please note I've been a programmer for a long time but I've few experience with C/CPP.

I hope someone can shed some light into this.
I believe that for someone experienced it may be a simple thing - if it's not I can make the necessary changes to the sample programs and upload them for testing, as I managed to replicate it by slightly altering the original samples.

Thank you for your time!

PLEASE NOTE: I'm only posting here because I do believe this is a bug but unfortunately I've not managed to track it yet, however I'll keep trying and keep this posted.

**OS**  
OSX

**mbed TLS build:**  
Version: 66b7edb10865809a972362ce362c62820075ade3
OS version: 10.14.4
Configuration: [used_config.txt](https://github.com/ARMmbed/mbedtls/files/3318655/used_config.txt)

  
Compiler and options (if you used a pre-built binary, please indicate how you obtained it):  
I'm using make with a simple config to define the config path.

**Expected behavior**   
Handshake should work on every new connection.

**Actual behavior**  
Handshake is only working every 2 connections.
There's no error on the server side, just a timeout on the client side and consecutive retries - all end up in error.

**Steps to reproduce**
Using the demos at programs/ssl:  
- Add fork to the DTLS server on every successfully connection (after handshake).
- Set a loop on writing and reading both on the DTLS server and client so that the sockets stay on.
- Start the server.
- Connect a first client (it will work).
- Connect a 2nd client, bad.txt will happen and it will get stuck on the handshake. 
- Connect a 3rd client and it will work!
- Connect a 4th and it will get stuck on the handshake.
- Connect a 5th and it will work.
And so on....

**Comment by RonEld** [comment]
Hi @WhiteaglePT 
Have you tried using the [`ssl_fork_server`](https://github.com/ARMmbed/mbedtls/blob/development/programs/ssl/ssl_fork_server.c) sample application with the `ssl_client2` application? I would try first doing a TLS handshake, and then modify the server to do a DTLS handshake.''

The log is not entirely clear, but could you check if you are using a seperate `mbedtls_ssl_context` for every fork \ connection?

**Comment by WhiteaglePT** [comment]
Hi!
Thank you for your reply!

I based my code on top of DTLS client & server sample applications.

As per your request I made sure to double check my resets properly and not logs are fine and back to normal on both versions!

The error persists,

However I came across to issue #1124 and by setting mbedtls_ssl_cookie_write and mbedtls_ssl_cookie_check to NULL on mbedtls_ssl_conf_dtls_cookies it started working - so the error is somewhere on HelloVerify.

I'm closing this as I'm sure it's not library related at this point - even though I'm resetting both the cache context & cookie context I'm sure it's something I'm missing here.

Thanks again for getting me unstuck!
